### PR TITLE
Fix #568

### LINKED ISSUE #568: GraphiQL fetcher uses HTTP for subscriptions
I have been experimenting with subscriptions and the way mercurius currently uses GraphiQL is slightly wrong. The [fetcher](https://github.com/mercurius-js/mercurius/blob/e0f672c4a3dc2054e9557d863183aa58e265f34b/static/main.js#L21-L30) always uses the fetch API which should not be used for subscriptions which use WebSockets.

Apparently, there is a package (`@graphiql/create-fetcher`) which comes with a function that allows for an easy way to create a fetcher, there is some good documentation on this here: https://github.com/graphql/graphiql/blob/main/packages/graphiql-toolkit/docs/create-fetcher.md

Do you agree with me that the current way of doing it is wrong? All subscriptions just return `null` when requested via GraphiQL or even throw an error if the response is non-nullable.

I can maybe look into making a PR about this next week if we agree on using `@graphiql/create-fetcher`.

### PR DESCRIPTION
**Pull request by puchm**
Fix for https://github.com/mercurius-js/mercurius/issues/568
This is a draft.

I have tried using `GraphiQL.createFetcher` in order to create a fetcher that works properly for subscriptions.

However, the WebSocket is getting closed with code **1002** every time I try to open a subscription. This has already been reported in https://github.com/graphql/graphiql/issues/1800 but the issue is still the same after changing the GraphiQL version.

Any ideas?

### CODE REVIEW

**Comment by mcollina** [comment]
I don't know why it fails :( Have you tried looking at the messages that are being sent back and forth?

What's the server are you testing against?

**Comment by puchm** [comment]
@mcollina I am testing against a simple mercurius server.
<details>
<summary>Server implementation</summary>

```ts
const Fastify = require('fastify');
const mercurius = require('./mercurius');

const app = Fastify();

const gql = String.raw

const schema = gql`
    type Query {
        me: String
    }

    type Subscription {
        commentAdded: Comment!
    }

    type Comment {
        text: String
    }

    type Mutation {
        addComment(text: String): Boolean!
    }
`

const resolvers = {
    Query: {
        me: () => 'Hello there.'
    },
    Mutation: {
        addComment: (_parent, { text }, { pubsub }) => {
            pubsub.publish({
                topic: 'COMMENT',
                payload: {
                    value: text
                }
            });

            return true;
        }
    },
    Subscription: {
        commentAdded: {
            subscribe: (_root, _args, { pubsub }) => pubsub.subscribe('COMMENT')
        }
    }
}

app.register(mercurius, {
    graphiql: true,
    resolvers: resolvers,
    schema: schema,
    subscription: {
        onConnect: () => {
            console.log("hiii")
        }
    }
})

app.listen(4001).then(() => console.log("HI there"))
```
</details>

I asked on the Apollo GraphQL discord and apparently this could be caused by the server using the legacy websockets protocol instead of `graphql-ws`.

**Comment by puchm** [comment]
It is working now using the `legacyClient` option. Request headers and query variables are also working fine.

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
> It is working now using the `legacyClient` option. Request headers and query variables are also working fine.

Good spot!

**Comment by puchm** [comment]
Good point, fixed that.

File: static/main.js [hunk 1]
```diff
@@ -17,20 +17,31 @@ const importer = {
...
 
-    return res.json()
-  }
+  //   return res.json()
+  // }
+
+  const host = window.location.host
+
+  const url = `http://${host}${window.GRAPHQL_ENDPOINT}`
+  const subscriptionUrl = `ws://${host}${window.GRAPHQL_ENDPOINT}`
```
[the remark this answers is not in the document]
Comment (mcollina) [reply static/main.js:23]: You need to use `wss` when `window.location.protocol` is `https`.

**Comment by puchm** [comment]
Should be fine now.

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**